### PR TITLE
Problem: memiavl's memory copy behavior is not consistent

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #1009
- make the default behavior safe
- provide the unsafe zero-copy option

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/db_test.go [hunk 1]
```diff
@@ -233,3 +235,32 @@ func TestLoadVersion(t *testing.T) {
...
+	db.SetZeroCopy(false)
+	valueCloned := db.TreeByName("test").Get([]byte("hello"))
+	require.Equal(t, []byte("world"), valueCloned)
+
+	require.NoError(t, db.Close())
+
+	require.Equal(t, []byte("world"), valueCloned)
+
+	// accessing the zero-copy value after the db is closed triggers segment fault.
+	debug.SetPanicOnFault(true)
```
Comment (yihuang) [inline memiavl/db_test.go:262]:
```suggestion
	defer debug.SetPanicOnFault(debug.SetPanicOnFault(true))
```

File: memiavl/tree.go [hunk 2]
```diff
@@ -141,6 +154,9 @@ func (t *Tree) GetWithIndex(key []byte) (int64, []byte) {
 	}
 
 	value, index := t.root.Get(key)
+	if !t.zeroCopy {
+		value = bytes.Clone(value)
+	}
 	return int64(index), value
```
Comment (mmsqe) [inline memiavl/tree.go:160]: node is return [uint32](https://github.com/crypto-org-chain/cronos/blob/37c5710e933c93cdda82753a9cba285d69f14e30/memiavl/node.go#L26), why we need convert index to int64

File: memiavl/db_test.go [hunk 3]
```diff
@@ -233,3 +235,32 @@ func TestLoadVersion(t *testing.T) {
...
+	db.SetZeroCopy(false)
+	valueCloned := db.TreeByName("test").Get([]byte("hello"))
+	require.Equal(t, []byte("world"), valueCloned)
+
+	require.NoError(t, db.Close())
+
+	require.Equal(t, []byte("world"), valueCloned)
+
+	// accessing the zero-copy value after the db is closed triggers segment fault.
+	defer debug.SetPanicOnFault(debug.SetPanicOnFault(true))
```
Comment (mmsqe) [inline memiavl/db_test.go:262]:
```suggestion
	// reset global panic on fault setting after function finished
	defer debug.SetPanicOnFault(debug.SetPanicOnFault(true))
```

File: memiavl/tree.go (hunk 2 above)
Comment (yihuang) [reply memiavl/tree.go:160]: The public API use int64, to keep it similar with existing iavl

**Review by mmsqe — APPROVED** [review]

**Pull request merged**